### ISSUE TITLE
Kotlin serialization ignored when produce application/x-ndjson

### ISSUE THREAD

**Issue by Distractic** [spring-projects/spring-framework#30398]
**Affects:**  3.0.6

---

I want to make a REST API using Spring and Kotlin.

```kt
plugins {
    id("org.springframework.boot") version "3.0.6"
    id("io.spring.dependency-management") version "1.1.0"
    kotlin("jvm") version "1.8.21"
    kotlin("plugin.serialization") version "1.8.21"
    kotlin("plugin.spring") version "1.8.21"
}
```

I have this data class, I want to serialize it using [Kotlin serialization](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/json.md)

```kt
@Serializable
data class Account(
    var name: String,
    var email: String,
    var role: String,
    @SerialName("is_active")
    var active: Boolean,
    var id: String,
    @SerialName("created_at")
    @Serializable(with = InstantSerializer::class)
    var createdAt: Instant = Instant.EPOCH
)
```

This is  my controller that returns an `application/json` content :

```kt
@RestController
@RequestMapping("/accountExamples")
class AccountExampleController {

    @GetMapping
    fun getAll(): Flow<AccountPojo> = List(2) {
        AccountPojo(
            name = "name $it",
            email = "email $it",
            role = "role $it",
            active = true
        )
    }.asFlow()

}
```

When I call it, everything is correct : 
```json
[
    {
        "name": "name 0",
        "email": "email 0",
        "role": "role 0",
        "is_active": true,
        "id": "0",
        "created_at": "2023-04-29T09:34:03.103Z"
    },
    {
        "name": "name 1",
        "email": "email 1",
        "role": "role 1",
        "is_active": true,
        "id": "1",
        "created_at": "2023-04-29T09:34:03.103Z"
    }
]
```

However, when I want to produce an `application/x-ndjson` result using  :

```kt
@RestController
@RequestMapping("/accountExamples")
class AccountExampleController {

    @GetMapping(produces = [APPLICATION_NDJSON_VALUE])
    fun getAll(): Flow<AccountPojo> = List(2) {
        AccountPojo(
            name = "name $it",
            email = "email $it",
            role = "role $it",
            active = true,
            id = "$it",
            createdAt = Instant.now()
        )
    }.asFlow()

}
```

This  is the json result :  
```json
{
    "name": "name 0",
    "email": "email 0",
    "role": "role 0",
    "active": true,
    "id": "0",
    "createdAt": "2023-04-29T09:35:12.328055700Z"
}
{
    "name": "name 1",
    "email": "email 1",
    "role": "role 1",
    "active": true,
    "id": "1",
    "createdAt": "2023-04-29T09:35:12.328055700Z"
}
```

As you can see, [Kotlin serialization](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/json.md) is not used to serialize objects when I want to produce `application/x-ndjson`. Instead of it, Jackson is used.

In this case, if we set the `JsonProperty` annotation on a field, the final json will be modified

```kt
@Serializable
data class Account(
    var name: String,
    var email: String,
    var role: String,
    @SerialName("is_active")
    var active: Boolean,
    var id: String,
    @SerialName("created_at")
    @JsonProperty("test") // Jackson annotation here
    @Serializable(with = InstantSerializer::class)
    var createdAt: Instant = Instant.EPOCH
)
```

Result :

```json
{
    "test": "2023-04-29T09:38:31.009048100Z",
    "name": "name 0",
    "email": "email 0",
    "role": "role 0",
    "active": true,
    "id": "0"
}
{
    "test": "2023-04-29T09:38:31.009048100Z",
    "name": "name 1",
    "email": "email 1",
    "role": "role 1",
    "active": true,
    "id": "1"
}
```

**Comment by sdeleuze** [comment]
Hi, the thing is that Kotlin Serialization support in Spring does not support yet streaming. If you use case is just about finite `Flow` serialization, could you please try to extend Spring codec/converter to specify customized mime type? Or do you need streaming support?

**Comment by Distractic** [comment]
I don't need streaming support for the moment. I was just curious about the use of Kotlin in Spring so I tried it. In my test, I used the maximum of Kotlin component (serializable, coroutine etc.)
So, when I see this feature is missing, I created an issue to alert about it because I would like use every Kotlin tools (and Spring) to create a server.
Do you think the support of streaming can be implemented in Spring in the future?

**Comment by sdeleuze** [comment]
It will require Kotlin/kotlinx.serialization#253 to be fixed, so let's keep this issue opened but blocked on this issue.

**Comment by sdeleuze** [comment]
Superseded by #32074.
